### PR TITLE
make k8s support cephfs fuse mount

### LINKED ISSUE #46636: FUSE client support for K8s CephFS
The goal of this issue is to design a CephFS driver using FUSE client for Kubernetes.</br>
Existing Ceph kernel client has the following drawbacks:</br>
•	Kernel client does not support client-enforced quotas.
•	Kernel CephFS doesn't support quota or capacity, capacity requested by PVC is not enforced or validated.
•	Kernel CephFS doesn't work with SELinux, setting SELinux label in Pod's security context will not work.

@kubernetes/sig-storage

### PR DESCRIPTION
**Pull request by zhangxiaoyu-zidif**
**What this PR does / why we need it**:
make  k8s support cephfs fuse.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/kubernetes/kubernetes/issues/46636, https://github.com/kubernetes/kubernetes/issues/44486

**Special notes for your reviewer**:

**Release note**:

```release-note
K8s supports cephfs fuse mount.
```

### CODE REVIEW

**Comment by zhangxiaoyu-zidif** [comment]
/cc @rootfs 
I create a new PR from https://github.com/kubernetes/kubernetes/pull/55222
Because #55222  has too many time rebase and update, I am afraid that I can squash it when merge...

**Comment by zhangxiaoyu-zidif** [comment]
@rootfs @cofyc 
we have tested on cephfs fuse.

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: ceph-secret
data:
  key: QVFCRHZNaFpmM3VMSEJBQStCK0FJUU1Yc005SzNXamVGM05JcGc9PQo=
```

```yaml
apiVersion: v1
kind: PersistentVolume
metadata:
  name: pv2
spec:
  capacity:
    storage: 10Gi
  accessModes:
    - ReadWriteMany
  cephfs:
    monitors:
    - 4.4.4.159:6789
    path:  /111
    user: admin
    secretRef:
      name: ceph-secret
  persistentVolumeReclaimPolicy: Retain
```

```yaml
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: pvc1
spec:
  accessModes:
      - ReadWriteMany
  resources:
      requests:
        storage: 8Gi
```

```yaml
kind: Pod
apiVersion: v1
metadata:
  name: pod1
spec:
  containers:
    - name: pod1
      image: gcr.io/google_containers/nginx:1.7.9
      env:
      - name: GET_HOSTS_FROM
        value: env
      ports:
      - containerPort: 81
      volumeMounts:
      - name: cephfs1
        mountPath: "/home/cephfs"
  automountServiceAccountToken: false
  volumes:
  - name: cephfs1
    persistentVolumeClaim:
      claimName: pvc1
```

File: pkg/volume/cephfs/cephfs.go [hunk 1]
```diff
@@ -231,15 +234,31 @@ func (cephfsVolume *cephfsMounter) SetUpAt(dir string, fsGroup *int64) error {
...
+		err = cephfsVolume.execFuseMount(dir)
+		if err != nil {
+			// cleanup upon failure
+			keyringPath := cephfsVolume.GetKeyringPath()
+			_, StatErr := os.Stat(keyringPath)
+			if !os.IsNotExist(StatErr) {
+				os.RemoveAll(keyringPath)
+			}
+			util.UnmountPath(dir, cephfsVolume.mounter)
+			return err
```
Comment (rootfs) [inline pkg/volume/cephfs/cephfs.go]: fall back to kernel cephfs

File: pkg/volume/cephfs/cephfs.go [hunk 2]
```diff
@@ -299,6 +330,90 @@ func (cephfsVolume *cephfs) execMount(mountpoint string) error {
...
+		fileProjection.Mode = int32(0644)
+		fileName := cephfsVolume.id + ".keyring"
+
+		payload[fileName] = fileProjection
+
+		writerContext := fmt.Sprintf("cephfuse:%v.keyring", cephfsVolume.id)
+		writer, err := util.NewAtomicWriter(keyringPath, writerContext)
+		if err != nil {
+			glog.Errorf("failed to create atomic writer: %v", err)
+			return err
```
Comment (rootfs) [inline pkg/volume/cephfs/cephfs.go]: cleanup

File: pkg/volume/cephfs/cephfs.go [hunk 3]
```diff
@@ -299,6 +330,90 @@ func (cephfsVolume *cephfs) execMount(mountpoint string) error {
...
+		}
+
+		err = writer.Write(payload)
+		if err != nil {
+			glog.Errorf("failed to write payload to dir: %v", err)
+			return err
```
Comment (rootfs) [inline pkg/volume/cephfs/cephfs.go]: cleanup

File: pkg/volume/cephfs/cephfs.go [hunk 4]
```diff
@@ -264,6 +287,14 @@ func (cephfsVolume *cephfs) GetPath() string {
 	return cephfsVolume.plugin.host.GetPodVolumeDir(cephfsVolume.podUID, utilstrings.EscapeQualifiedNameForDisk(name), cephfsVolume.volName)
 }
 
+// GateKeyringPath creates cephfuse keyring path
+func (cephfsVolume *cephfs) GetKeyringPath() string {
```
Comment (rootfs) [inline pkg/volume/cephfs/cephfs.go]: this leaks keyring - anybody has access to the directory can see the keyring. 

see https://github.com/kubernetes/kubernetes/blob/release-1.8/pkg/volume/rbd/rbd_util.go#L297

File: pkg/volume/cephfs/cephfs.go [hunk 5]
```diff
@@ -231,15 +234,31 @@ func (cephfsVolume *cephfsMounter) SetUpAt(dir string, fsGroup *int64) error {
...
+		glog.V(4).Infof("CephFS fuse mount")
+		err = cephfsVolume.execFuseMount(dir)
+		if err != nil {
+			// cleanup upon failure
+			keyringPath := cephfsVolume.GetKeyringPath()
+			_, StatErr := os.Stat(keyringPath)
+			if !os.IsNotExist(StatErr) {
+				os.RemoveAll(keyringPath)
+			}
+			util.UnmountPath(dir, cephfsVolume.mounter)
```
Comment (rootfs) [inline pkg/volume/cephfs/cephfs.go]: unmount if it is a mountpoint

File: pkg/volume/cephfs/cephfs.go [hunk 6]
```diff
@@ -255,6 +274,10 @@ func (cephfsVolume *cephfsUnmounter) TearDown() error {
 
 // TearDownAt unmounts the bind mount
 func (cephfsVolume *cephfsUnmounter) TearDownAt(dir string) error {
+	glog.V(4).Infof("TearDownAt: dir: %q ", dir)
+	if strings.Contains(dir, "~keyring") {
```
Comment (rootfs) [inline pkg/volume/cephfs/cephfs.go]: is the fs mounted here?

File: pkg/volume/cephfs/cephfs.go (hunk 3 above)
Comment (zhangxiaoyu-zidif) [reply pkg/volume/cephfs/cephfs.go]: the cleanup will be in https://github.com/kubernetes/kubernetes/pull/55866/files#diff-021042d70c9aac8994638a8d0827b1e0R242

File: pkg/volume/cephfs/cephfs.go (hunk 2 above)
Comment (zhangxiaoyu-zidif) [reply pkg/volume/cephfs/cephfs.go]: many thanks for review. the cleanup will be in https://github.com/kubernetes/kubernetes/pull/55866/files#diff-021042d70c9aac8994638a8d0827b1e0R242

File: pkg/volume/cephfs/cephfs.go (hunk 4 above)
Comment (zhangxiaoyu-zidif) [reply pkg/volume/cephfs/cephfs.go]: I change it, no matter if mount fail, we delete the keyring after execFuseMount

File: pkg/volume/cephfs/cephfs.go (hunk 6 above)
Comment (zhangxiaoyu-zidif) [reply pkg/volume/cephfs/cephfs.go]: I delete them here.

**Comment by zhangxiaoyu-zidif** [comment]
ping @rootfs , thanks you for review. I do some fixes. I really hope we can aboard on release-1.9.
My team is very interest in Cephfs fuse and use it in producive env, we will continue to focus on this thesis. Thanks for your help sincerally.

File: pkg/volume/cephfs/cephfs.go [hunk 7]
```diff
@@ -231,15 +234,32 @@ func (cephfsVolume *cephfsMounter) SetUpAt(dir string, fsGroup *int64) error {
...
+		glog.V(4).Infof("CephFS fuse mount.")
+		err = cephfsVolume.execFuseMount(dir)
+		// cleanup no matter if fuse mount fail.
+		keyringPath := cephfsVolume.GetKeyringPath()
+		_, StatErr := os.Stat(keyringPath)
+		if !os.IsNotExist(StatErr) {
+			os.RemoveAll(keyringPath)
+		}
+		if err == nil {
+			// cephfs fuse mount successed.
```
Comment (rootfs) [inline pkg/volume/cephfs/cephfs.go]: succeeded

File: pkg/volume/cephfs/cephfs.go [hunk 8]
```diff
@@ -299,6 +327,90 @@ func (cephfsVolume *cephfs) execMount(mountpoint string) error {
...
+		return false
+	}
+	return false
+}
+
+func (cephfsVolume *cephfs) execFuseMount(mountpoint string) error {
+	// cephfs keyring file
+	keyring_file := ""
+	// override secretfile if secret is provided
+	if cephfsVolume.secret != "" {
```
Comment (rootfs) [inline pkg/volume/cephfs/cephfs.go]: add a TODO here. cephfs fuse currently doesn't support secret option. Remove keyring file create once secret option is supported.

**Comment by rootfs** [comment]
@zhangxiaoyu-zidif almost there, ping me once you update. thanks

**Comment by zhangxiaoyu-zidif** [comment]
@rootfs  ,done. Thanks very much. 
We will feedback more features and infos about cephfs fuse.
Thanks for your help.

**Comment by rootfs** [comment]
@zhangxiaoyu-zidif thanks, squash commits and we are ready to go

**Comment by zhangxiaoyu-zidif** [comment]
@rootfs , squashed. Thanks!

**Comment by zhangxiaoyu-zidif** [comment]
@luxas , @rootfs   could you add it to release 1.9?

**Comment by zhangxiaoyu-zidif** [comment]
@ericchiang  for milestone 1.9

**Comment by zhangxiaoyu-zidif** [comment]
@idvoretskyi for 1.9 milestone. Thanks！

**Comment by ericchiang** [comment]
cc @kubernetes/sig-storage-feature-requests

**Comment by ericchiang** [comment]
cc @kubernetes/sig-storage-pr-reviews for milestone

**Pull request merged**